### PR TITLE
Fix(store) Fix 4450

### PR DESCRIPTION
**Pull request by ailisp**
Root cause: 48 values stored in TransactionResult column were not in correct form. This only happen in mainnet archive node. Possibly caused by previous migrations were broken. By manually inspection bytes and commit/migration history I found the TransactionResult was a single ExecutionOutcomeWithIdAndProof, changed to `vec[ExecutionOutcomeWithIdAndProof]` on commit: bf9ae4c, and these 48 incorrect ones are correct if interprete as single `ExecutionOutcomeWithIdAndProof`. But both the code and the migration introduced in that commit looks legit. Not sure how these several unmigrated outcomes sneaked in.

Therefore the fix is straightforward: borsh deserialize failed ones as single ExecutionOutcomeWithIdAndProof it success and without bytes left.

Test Plan
---------
Have tested with latest mainnet archive backup. migration is successful and node started. And CI.

### CODE REVIEW

File: nearcore/src/migrations.rs [hunk 1]
```diff
@@ -349,6 +354,87 @@ pub fn migrate_22_to_23(path: &String, near_config: &NearConfig) {
...
+    for (key, value) in store.iter(DBCol::ColTransactionResult) {
+        if Vec::<ExecutionOutcomeWithIdAndProof>::try_from_slice(&value).is_ok() {
+            // has success in previous attempt of this migration
+            continue;
+        }
+
+        let old_outcomes = Vec::<OldExecutionOutcomeWithIdAndProof>::try_from_slice(&value);
+        let old_outcomes = if old_outcomes.is_ok() {
+            old_outcomes.unwrap()
+        } else {
```
Comment (matklad) [inline nearcore/src/migrations.rs:422]: `if is_ok() / .unwrap()` is an anti-pattern, use match or if-let to statically guarantee that unwrap can't fire

File: nearcore/Cargo.toml [hunk 2]
```diff
@@ -25,6 +25,7 @@ tracing = "0.1.13"
 num-rational = { version = "0.3", features = ["serde"] }
 near-rust-allocator-proxy = "0.2.8"
 lazy-static-include = "3"
+bs58 = "0.4.0"
```
Comment (matklad) [inline nearcore/Cargo.toml:28]: Seems like we don't actually use bs58 in the code?

**Review by bowenwang1996 — COMMENTED** [review]
@khorolets please double check that this works with indexer

**Comment by ailisp** [comment]
@khorolets since the change is only from `an execution outcome` to `vec[an execution outcome]` on some values of ColTransactionResult column, without actual data change and you were not having problem to decode the without decode the vec, so I assume there is no update needed in indexer. If you do want to double check, here is all keys of the `ColTransactionResult`, base58 encoded, that has changed:

```
h6QEnvsiX5HcgEqKt5PNLKkm4ztZHM9BxYFmEhw6pxf
mupGNKYa7BGLaNhgQiEtrbN2y198aujLPZeJsirrWK3
rz4PbLxEdizwGUkvzxoPJm4YnA9MGHRuCwnPLU24LcA
yczJmX7eNisN4NpqWEkpV6aRNponGJZAHy7ADGfzAsA
2srYRfs8p5D8PX5uBcTbT2yGq6KX4YNo3pGLvANp9pnH
2vWaM9qVo7zzjf5yF1ytLBHBMj1wUb4AZANz1pV1cGa8
3YvdKGjqADbvqAZo6BNK1HJXsVt3a3VJdywhtLTGiSxY
3bpza4JysYmcbnRyhnyUbqh2QpwpjkyZEL3KqwaJfE31
3uuNjPr2PyitQHccPHjUMjTKL8cQKxmz2aVsEtbsYy4S
3w6kSmrG5i1URrqWqTVwnMqmqUZ8z5UfJZki4nBCiydA
3zodkcUqfFafaUrDHjDrSxUcg3AGrw5H3c4nsKRaj5oJ
4HoaDptWASX9JvdK2ppJ2vTYs9VYYpFkVuzBKjLMAQuq
4Z4h6sLaM8UXoUCWRuLXaXG67ZhKMGvKNcbpDjWLJPU5
4cVmAvDoK7x2dXwaYQW2PGVR1swcaHcC3SW9hJTgoEJj
5fLMyWVYqrZ9QwLJfb1S9LFVfQDjxqnsurvfukCYijbX
6Mktn8Y92CNXv8duGQXfXqMSWWr3FhXELruZSXECFZky
77NeeZsfdtRZzczg318mPJ4joLkGLHN4L25JwCzq5g9s
7AL2ELHdYR5U8j64o5TehkCry83j54vszaHZdLDVt843
7pzSvYZ19aecMjQt2Q8wQEMzaRMEA2irjhkTXaJmsspG
7sRXQ8uxb4VgDoU4ys9hoYUkyGuDf1QWTrw3tU7f6P7B
855sxTB9m6CfwuZ92mCg8BaZxv4KBvuky2PRhuRq4io5
8A4v6Z274cJsE3HioX7A5R8KeKY8MdC642hRdfZFBmWn
918bR3Gu1ugFtQ6732jaDqFKjg2K4HEkEt3WgfHSknVA
9Nj1U2dy4b8Q8EGhCBsBwCCicF97g1tH8QXf3RNETjk4
9kd86Hs615uZUqByrS38a4mX2h7n7NXYTwuAMG9HDkDY
B2J8WKb9bp99KP3FaVHQZx8C54y8cDqiqq42MRuvtHCF
B5fsEfiWsSYxsGDnznkDYGACyogvNv84PfZEdZncv9JL
BGkPJ3u8hQZrhZMtBE7pzHXqXjgXUV4VojrfGKcsYZTV
BJ2BNXscPJSfLAC43DX2gJoqjSnS41R4PBBCrhop2t9E
D6kyrrFLxrGHVh4De5nY5sPuyrcVVywEBjjHjYsaPV63
DMNViYUfU9FMrfkzJuEsdG4utNNLTT6L1PLrx979d47i
DuVZR11eBU11Pb2KxiMwtwxp6SZ9R1oPwwgQRM4nqqdA
Ep9LdNtorphvdBHpN9Tp9CgXcZC6HVDyoiiv4yXYsYDv
FCHBThVdtN6v4Bq7YdQcyL9Xkaj7P4jRT4zZYnXxYMVR
FU7vtGsqPNUni4fksWJP2aoXMxpLt8wyMPtzg9dkhjxh
Fr85sMRxj6vvriB7S9iQzpm9Cjh3TqWJsjHxSKag9keP
Fuwx3ANPXt3Dfnw8wnM9SfGYPxzmkewzJ9beXvmwdMeT
G2MCjxDQWAkHcy5TGzXzfxbr43NbPPocQJ8LTkddZYsY
GE5XJJVBCY2RQbMcMKWGAw934F2FffoHzCdYE98t7EPE
GJGSnMdCxSMtU2U8jmMsoCCEVUpYENc5puzvARDuJVi1
GdzCzx868dXbG5wtnswhjPjuTCz28yX2RCWgH8v5CEiE
GgXzr879Uib98S8cMbQR5PrVNE8RioUYctt9bPF4Z6ZX
GyVcpCuzS7q497sTx9z3R2GnfWFiVGtgz4QZm8frwdqx
H7AAsGeDMDiVduGt7MSKKNzQyfZm18Vw7YhyXoNTpE66
HBnG9EAZkgxok19qgeRMxYGdbaEkMDRnUzEf25BZ9LjV
HRaTedBJi16LnGHbwoqgtMex6B6ipXamFqo5BLnpgDZV
HzrQVtMnHbKVWS8uRHCgsXnWUsN47mHfMrEZ43uw6X21
JDnzVU3qQpKGdwxuKnzfqJv7EsoawPh9PQBH8ZwU2BPE
```

Each key means the ExecutionOutcome id, and according to nearcore doc, they are either the receipt id or the transaction id that generates this execution outcome. Note if it's a local receipt then it's not stored in the store.

**Comment by khorolets** [comment]
@ailisp any at least approximate range of blocks where those are? I want to check, quick check showed I don't have them in indexer database, and we do store local receipts there. So I need to check it with indexer like Bowen suggested.

**Comment by ailisp** [comment]
@khorolets Here is the block hash of all affected execution outcomes:
```
outcome id: h6QEnvsiX5HcgEqKt5PNLKkm4ztZHM9BxYFmEhw6pxf, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX                                                                   
outcome id: mupGNKYa7BGLaNhgQiEtrbN2y198aujLPZeJsirrWK3, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX                                                                   
outcome id: rz4PbLxEdizwGUkvzxoPJm4YnA9MGHRuCwnPLU24LcA, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX                                                                   
outcome id: yczJmX7eNisN4NpqWEkpV6aRNponGJZAHy7ADGfzAsA, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX                                                                   
outcome id: 2srYRfs8p5D8PX5uBcTbT2yGq6KX4YNo3pGLvANp9pnH, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 2vWaM9qVo7zzjf5yF1ytLBHBMj1wUb4AZANz1pV1cGa8, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 3YvdKGjqADbvqAZo6BNK1HJXsVt3a3VJdywhtLTGiSxY, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 3bpza4JysYmcbnRyhnyUbqh2QpwpjkyZEL3KqwaJfE31, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 3uuNjPr2PyitQHccPHjUMjTKL8cQKxmz2aVsEtbsYy4S, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 3w6kSmrG5i1URrqWqTVwnMqmqUZ8z5UfJZki4nBCiydA, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 3zodkcUqfFafaUrDHjDrSxUcg3AGrw5H3c4nsKRaj5oJ, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 4HoaDptWASX9JvdK2ppJ2vTYs9VYYpFkVuzBKjLMAQuq, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 4Z4h6sLaM8UXoUCWRuLXaXG67ZhKMGvKNcbpDjWLJPU5, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 4cVmAvDoK7x2dXwaYQW2PGVR1swcaHcC3SW9hJTgoEJj, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 5fLMyWVYqrZ9QwLJfb1S9LFVfQDjxqnsurvfukCYijbX, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 6Mktn8Y92CNXv8duGQXfXqMSWWr3FhXELruZSXECFZky, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 77NeeZsfdtRZzczg318mPJ4joLkGLHN4L25JwCzq5g9s, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 7AL2ELHdYR5U8j64o5TehkCry83j54vszaHZdLDVt843, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 7pzSvYZ19aecMjQt2Q8wQEMzaRMEA2irjhkTXaJmsspG, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 7sRXQ8uxb4VgDoU4ys9hoYUkyGuDf1QWTrw3tU7f6P7B, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 855sxTB9m6CfwuZ92mCg8BaZxv4KBvuky2PRhuRq4io5, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 8A4v6Z274cJsE3HioX7A5R8KeKY8MdC642hRdfZFBmWn, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 918bR3Gu1ugFtQ6732jaDqFKjg2K4HEkEt3WgfHSknVA, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 9Nj1U2dy4b8Q8EGhCBsBwCCicF97g1tH8QXf3RNETjk4, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: 9kd86Hs615uZUqByrS38a4mX2h7n7NXYTwuAMG9HDkDY, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: B2J8WKb9bp99KP3FaVHQZx8C54y8cDqiqq42MRuvtHCF, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: B5fsEfiWsSYxsGDnznkDYGACyogvNv84PfZEdZncv9JL, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: BGkPJ3u8hQZrhZMtBE7pzHXqXjgXUV4VojrfGKcsYZTV, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: BJ2BNXscPJSfLAC43DX2gJoqjSnS41R4PBBCrhop2t9E, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: D6kyrrFLxrGHVh4De5nY5sPuyrcVVywEBjjHjYsaPV63, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: DMNViYUfU9FMrfkzJuEsdG4utNNLTT6L1PLrx979d47i, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: DuVZR11eBU11Pb2KxiMwtwxp6SZ9R1oPwwgQRM4nqqdA, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: Ep9LdNtorphvdBHpN9Tp9CgXcZC6HVDyoiiv4yXYsYDv, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: FCHBThVdtN6v4Bq7YdQcyL9Xkaj7P4jRT4zZYnXxYMVR, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: FU7vtGsqPNUni4fksWJP2aoXMxpLt8wyMPtzg9dkhjxh, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: Fr85sMRxj6vvriB7S9iQzpm9Cjh3TqWJsjHxSKag9keP, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: Fuwx3ANPXt3Dfnw8wnM9SfGYPxzmkewzJ9beXvmwdMeT, block hash: Ct1hVhJ1LtQ1iD7qnE6HVVkoVf9w4sDh5dNJKGNYtvqG
outcome id: G2MCjxDQWAkHcy5TGzXzfxbr43NbPPocQJ8LTkddZYsY, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: GE5XJJVBCY2RQbMcMKWGAw934F2FffoHzCdYE98t7EPE, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: GJGSnMdCxSMtU2U8jmMsoCCEVUpYENc5puzvARDuJVi1, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: GdzCzx868dXbG5wtnswhjPjuTCz28yX2RCWgH8v5CEiE, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: GgXzr879Uib98S8cMbQR5PrVNE8RioUYctt9bPF4Z6ZX, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: GyVcpCuzS7q497sTx9z3R2GnfWFiVGtgz4QZm8frwdqx, block hash: EYjiKNT8g3xs5ChGZj32LFnBWJWe2JXB6tktmTjWZsvq
outcome id: H7AAsGeDMDiVduGt7MSKKNzQyfZm18Vw7YhyXoNTpE66, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: HBnG9EAZkgxok19qgeRMxYGdbaEkMDRnUzEf25BZ9LjV, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: HRaTedBJi16LnGHbwoqgtMex6B6ipXamFqo5BLnpgDZV, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: HzrQVtMnHbKVWS8uRHCgsXnWUsN47mHfMrEZ43uw6X21, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
outcome id: JDnzVU3qQpKGdwxuKnzfqJv7EsoawPh9PQBH8ZwU2BPE, block hash: wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX
```

File: core/store/Cargo.toml [hunk 3]
```diff
@@ -21,6 +21,7 @@ fs2 = "0.4"
 tracing = "0.1"
 borsh = "0.8.1"
 thiserror = "1"
+bs58 = "0.4.0"
```
Comment (matklad) [inline core/store/Cargo.toml:24]: Seems like we don't need this either

**Comment by khorolets** [comment]
@ailisp 

So I've build the indexer with your changes. I checked archival rpc the block `wuDB3jtViUR94HsepyA3ifQXcWhYJFoAfNUSKNDDNJX` it has the height `20004768`

Checking the indexer database showed that we don't have this block.

I've started the indexer from block `20004767` and I get ...67 and then ...68

In indexer we're getting block by height via `ViewClient`, that's what I'm getting trying to get ...68

```
UnknownBlock { error_message: "BLOCK HEIGHT: 20004768" }
```
(It's essentially `DbNotFoudErr`)

Also, querying through JSON RPC (`ViewClient`'s `GetBlock`):

- Able to get the block by hash
- Unable to get the same block by height

I believe `store.get_block_hash_by_height()` doesn't have necessary references or something, I'm not an expert on that depth :(

Overall: I can't confirm indexer is fine here, because we've skipping the blocks which actually exist.

Anything else I can debug for you to help to resolve it?

**Comment by ailisp** [comment]
@khorolets Your steps look neat and correct. That's probably fine, actually, I found the outcome -> vec[outcome] happened at this https://github.com/near/nearcore/commit/bf9ae4ce8c680d3408db1935ebd0ca24c4960884, but both the code and the migration introduced in this commit looks legit. Not sure how these unmigrated things sneaked in. I have some potential guesses, but they don't quite explain why these blocks are gone
- there's pending db inserts of old format ExecutionOutcomeWithIdAndProof right before running the migration, cause the several old ones escaping the migration.
- Outcome of non-canonical chain? 
- state synced with other non migrated node?
- not properly run migration?

@bowenwang1996

**Comment by frol** [comment]
@ailisp @bowenwang1996 I believe those are blocks/outcomes of a non-canonical chain. Does GC clean up non-canonical chain data on archival nodes?

**Comment by ailisp** [comment]
@frol Very likely! GC does clean up outcomes on the given (non canonical) block on non archival mode, in chain/chain/src/store.rs `gc_outcomes`, which is triggered by chain.clear_data -> clear_block_data. However clear_data is called only on non archival node: `if !self.config.archive ... clear_data(` in chain/chain/src/store.rs it's the only call to clear_data.

So there should be much more transaction outcomes from non canonical chains then, We should clean them up in a separate PR that has cleanup migration and also implement periodic cleanup of non canonical block outcomes. Also, is there a use case user would ever query transaction outcomes from non canonical chains?

**Comment by frol** [comment]
@ailisp could you confirm that all those blocks are actually not part of the canonical change? That would be helpful to make the decision.

The code looks fine to me, but overall unnecessary, so I defer the decision whether we want to merge it or fix GC to @bowenwang1996.

>  Also, is there a use case user would ever query transaction outcomes from non canonical chains?

I would even say it is dangerous and we should not let those be queried.

**Comment by ailisp** [comment]
> The code looks fine to me, but overall unnecessary, so I defer the decision whether we want to merge it or fix GC to @bowenwang1996.

@frol it is blocking #4438, so i think it's better to have this quick fix before come up with a thorough plan on archive node gc. These incorrect format outcomes are lingering in mainnet archival node anyway and no harm to migrate them into the correct form first

**Comment by frol** [comment]
> Otherwise maybe the design choice was archival node keep more data and let indexer decide canonical? Which i think it's bad

I am not sure I understood you correctly. Nevertheless, Indexer only tracks finalized blocks and it does not go into the weeds of non-finalized blocks.

**Review by EgorKulikov — APPROVED** [review]
lgtm as to runtime changes

**Comment by bowenwang1996** [comment]
> I believe those are blocks/outcomes of a non-canonical chain. Does GC clean up non-canonical chain data on archival nodes?

No there is currently no GC for archival nodes.

**Review by bowenwang1996 — COMMENTED** [review]
It looks like we only need to run this migration for archival nodes.

**Comment by ailisp** [comment]
> I am not sure I understood you correctly. Nevertheless, Indexer only tracks finalized blocks and it does not go into the weeds of non-finalized blocks.

I mean I see archival node don't clean up outcomes from non-canonical chain in store. I'm not sure about the detail of indexer, but if it do receive this outcomes, it needs to tell whether it's outcome of a receipt in a canonical block itself. Which is a design choice to simplifies gc on archival node

**Comment by ailisp** [comment]
> It looks like we only need to run this migration for archival nodes.

Double checked migration should work on both archival and non archival node. And we do need to run migration on non archival node too - if user rpc query from non archival node, deserialize outcome would fail without the migration. Although after a few epoch, old outcomes would be cleanup but it is problemtic in these epochs.

**Review by bowenwang1996 — APPROVED** [review]

**Pull request merged**